server/src/main/java/org/elasticsearch/TransportVersion.java

```diff
@@ -0,0 +1,292 @@
+                        : "expected [" + version.id + "] to be uniquely mapped but saw [" + maybePrevious + "] and [" + version + "]";
+
+                    TransportVersion sameUniqueId = uniqueIds.put(version.uniqueId, version);
+                    assert sameUniqueId == null
+                        : "Versions "
+                            + version
+                            + " and "
+                            + sameUniqueId
+                            + " have the same unique id. Each TransportVersion should have a different unique id";
+
+                } catch (IllegalAccessException e) {
+                    assert false : "Version field [" + fieldName + "] should be public";
+                }
+            }
+        }
+        idToVersion = Map.copyOf(builder);
+        declaredVersions = idToVersion.values().stream().collect(Collectors.toCollection(() -> new TreeSet<>()));
+    }
+
+    public static TransportVersion readVersion(StreamInput in) throws IOException {
+        return fromId(in.readVInt());
+    }
+
+    public static TransportVersion fromId(int id) {
+        TransportVersion known = idToVersion.get(id);
+        if (known != null) {
+            return known;
+        }
+        // this is a version we don't otherwise know about - just create a placeholder
+        return new TransportVersion(id, "<unknown>");
+    }
+
+    public static void writeVersion(TransportVersion version, StreamOutput out) throws IOException {
+        out.writeVInt(version.id);
+    }
+
+    /**
+     * Returns the minimum version between the 2.
+     */
+    public static TransportVersion min(TransportVersion version1, TransportVersion version2) {
+        return version1.id < version2.id ? version1 : version2;
+    }
+
+    /**
+     * Returns the maximum version between the 2
+     */
+    public static TransportVersion max(TransportVersion version1, TransportVersion version2) {
+        return version1.id > version2.id ? version1 : version2;
+    }
+
+    // TODO for testing or a getter? should this be used?
+    final int id;
+    private final String uniqueId;
+
+    TransportVersion(int id, String uniqueId) {
+        this.id = id;
+        this.uniqueId = Strings.requireNonEmpty(uniqueId, "Each TransportVersion needs a unique string id");
+    }
+
+    public boolean after(TransportVersion version) {
+        return version.id < id;
+    }
+
+    public boolean onOrAfter(TransportVersion version) {
+        return version.id <= id;
+    }
+
+    public boolean before(TransportVersion version) {
+        return version.id > id;
+    }
+
+    public boolean onOrBefore(TransportVersion version) {
+        return version.id >= id;
+    }
+
+    @Override
+    public int compareTo(TransportVersion other) {
+        return Integer.compare(this.id, other.id);
+    }
+
+    /**
+     * returns a sorted list of declared transport version constants
+     */
+    public static NavigableSet<TransportVersion> getDeclaredVersions() {
```

Comment:
The name "declared" versions is a little confusing since there are not some other set of versions (the original name in Version comes from the class semantics of declared variables). Instead I would call this `getAllVersions()`.

Reply:
Updated